File: app/src/main/java/com/mmjang/ankihelper/data/dict/FormsUtil.java
package com.mmjang.ankihelper.data.dict;

import android.content.Context;

import com.mmjang.ankihelper.data.dict.FormsDatabase; // Import FormsDatabase
import com.mmjang.ankihelper.data.dict.Form; // Import Form
import com.mmjang.ankihelper.data.dict.FormsDao; // Import FormsDao

public class FormsUtil {

    private static FormsUtil instance = null;
    private FormsDatabase db; // Holds the Room database instance

    protected FormsUtil(Context context){
        // Initialize Room database
        db = FormsDatabase.getInstance(context);
    }

    public static FormsUtil getInstance(Context context){
        if(instance == null){
            instance = new FormsUtil(context);
        }
        return instance;
    }

    public String[] getForms(String q) {
        // Get forms from Room database
        String bases = db.formsDao().getForms(q.toLowerCase());

        if(bases == null || bases.isEmpty()){
            return new String[0];
        }
        return bases.split("@@@");
    }
}

File: app/src/main/java/com/mmjang/ankihelper/data/dict/FormsUtil.kt
package com.mmjang.ankihelper.data.dict

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class FormsUtil private constructor(context: Context) {
    private val db: FormsDatabase = FormsDatabase.getInstance(context)

    companion object {
        @Volatile
        private var instance: FormsUtil? = null

        fun getInstance(context: Context): FormsUtil {
            return instance ?: synchronized(this) {
                instance ?: FormsUtil(context).also { instance = it }
            }
        }
    }

    /**
     * Get word forms using suspend function for async operation
     * @param q The word to query
     * @return Array of base forms
     */
    suspend fun getForms(q: String): Array<String> = withContext(Dispatchers.IO) {
        try {
            val bases = db.formsDao().getForms(q.lowercase())

            if (bases.isNullOrEmpty()) {
                emptyArray()
            } else {
                bases.split("@@@").toTypedArray()
            }
        } catch (e: Exception) {
            // Log error and return empty array
            android.util.Log.e("FormsUtil", "Error getting forms for '$q'", e)
            emptyArray()
        }
    }

    /**
     * Synchronous wrapper for Java compatibility (uses runBlocking - use sparingly)
     * Prefer using the suspend version directly from Kotlin code
     */
    @Deprecated("Use suspend version instead", ReplaceWith("getForms(q)"))
    fun getFormsSync(q: String): Array<String> {
        return kotlinx.coroutines.runBlocking {
            getForms(q)
        }
    }
}
